R/calcHDDCDD.R

```diff
@@ -1,101 +1,714 @@
-#' HDD and CDD scenario and historic data
+#' calculate HDD and CDD based on outdoor/indoor temperature difference
 #'
-#' smoothes the HDD and CDD scenario data
+#' @description heating and cooling degree days based on raw outside temperature
+#'   or bais-adjusted internal temperature (BAIT), driver for space heating and
+#'   cooling demand in buildings
 #'
-#' @param tlimit Temperature threshold for computing HDD and CDD
+#' @param mappingFile data.frame containing input data file names and directories
+#' @param bait specify use of raw temperature or BAIT
+#' @param multiscen specify if mappingFile covers more than one scenario
 #'
-#' @author Antoine Levesque
+#' @return magpie object of heating and cooling degree days
 #'
-#' @importFrom madrat readSource calcOutput
-#' @importFrom magclass setNames setYears as.magpie getNames lowpass getYears mbind
-#' @importFrom dplyr %>%
-#' @importFrom quitte as.quitte
-#' @importFrom utils tail
-#' @export
-
-calcHDDCDD <- function(tlimit = 18) {
-
-  stopifnot(as.character(tlimit) %in% as.character(17:25))
-
-  data <- readSource("HDDCDD", subtype = tlimit)
-
-  # yearsScenario = all years for which history is either NA or 0
-  index <- apply(data[, , paste0("history.HDD.history.", tlimit)], c(2, 3),
-                 function(x) {
-                   return(all(is.na(x) | x == 0))
-                 })
-
-  yearsScenario <- c(gsub("y", "", attributes(index)$dimnames$period[index]) %>% as.integer(), 2100)
-  yearsHistory <- setdiff(getYears(data, as.integer = TRUE), yearsScenario)
-
-  convergenceYear <- 2030
-  convergenceYears <- c(max(yearsHistory), yearsScenario[yearsScenario <= convergenceYear])
-  afterConvergenceYears <- yearsScenario[!yearsScenario %in% convergenceYears]
-  lambda <- c(rep(1, length(yearsHistory)),
-              tail(seq(1, 0, length.out = length(convergenceYears)), -1),
-              rep(0, length(afterConvergenceYears)))
-  names(lambda) <- c(yearsHistory, yearsScenario)
-  lambda <- as.magpie(lambda)
-
-  rcps     <- grep("history", getNames(data, TRUE)$rcp,      value = TRUE, invert = TRUE)
-  ssps     <- grep("history", getNames(data, TRUE)$scenario, value = TRUE, invert = TRUE)
-  variable <- grep("history", getNames(data, TRUE)$variable, value = TRUE, invert = TRUE)
-
-  data[is.na(data)] <- 0
-  data <- mbind(data, (1 / 3) * (setYears(data[, 2099, ], 2100) +
-                                 setYears(data[, 2098, ], 2100) +
-                                 setYears(data[, 2097, ], 2100))) # give a point for 2100
-
-  # distribute the history between all scenarios
-  data <- do.call("mbind", lapply(rcps, function(rcp) {
-    do.call("mbind", lapply(ssps, function(ssp) {
-      tmp <- data[, , rcp][, , ssp]
-      tmp[, yearsHistory, ] <- setNames(data[, yearsHistory, "history"][, , "history"],
-                                        getNames(tmp[, yearsHistory, ]))
+#' @author Robin Krekeler, Hagen Tockhorn
```

Comment:
My fault, I know. My old name still haunting some functions...
```suggestion
#' @author Robin Hasse, Hagen Tockhorn
```